mobile/src/main/java/org/openhab/habdroid/ui/WriteTagActivity.java

```diff
@@ -139,76 +150,126 @@ public void onNewIntent(Intent intent) {
-            if (!sitemapUri.getPath().startsWith("/rest/sitemaps")) {
-                throw new URISyntaxException(mSitemapPage, "Expected a sitemap URL");
-            }
-            StringBuilder uriToWrite = new StringBuilder("openhab://sitemaps");
-            uriToWrite.append(sitemapUri.getPath().substring(14));
-            if (!TextUtils.isEmpty(mItem) && !TextUtils.isEmpty(mCommand)) {
-                uriToWrite.append("?item=").append(mItem).append("&command=").append(mCommand);
-            }
-            writeTagMessage.setText(R.string.info_write_tag_progress);
-            writeTag(tagFromIntent, uriToWrite.toString());
-        } catch (URISyntaxException e) {
-            Log.e(TAG, e.getMessage());
-            writeTagMessage.setText(R.string.info_write_failed);
-        }
-    }
+                URI sitemapUri;
+                String longUri = "";
+                String shortUri = "";
 
-    private void writeTag(Tag tag, String uri) {
-        Log.d(TAG, "Creating tag object for URI " + uri);
-        TextView writeTagMessage = findViewById(R.id.write_tag_message);
-
-        NdefRecord[] ndefRecords = new NdefRecord[] { NdefRecord.createUri(uri) };
-        NdefMessage message = new NdefMessage(ndefRecords);
-        NdefFormatable ndefFormatable = NdefFormatable.get(tag);
-
-        if (ndefFormatable != null) {
-            Log.d(TAG, "Tag is uninitialized, formating");
-            try {
-                ndefFormatable.connect();
-                ndefFormatable.format(message);
-                ndefFormatable.close();
-                writeTagMessage.setText(R.string.info_write_tag_finished);
-                autoCloseActivity();
-            } catch (IOException | FormatException e) {
-                Log.e(TAG, "Writing to unformatted tag failed: " + e);
-                writeTagMessage.setText(R.string.info_write_failed);
-            }
-        } else {
-            Log.d(TAG, "Tag is initialized, writing");
-            Ndef ndef = Ndef.get(tag);
-            if (ndef != null) {
                 try {
-                    Log.d(TAG, "Connecting");
-                    ndef.connect();
-                    Log.d(TAG, "Writing");
-                    if (ndef.isWritable()) {
-                        ndef.writeNdefMessage(message);
+                    sitemapUri = new URI(mSitemapPage);
+                    if (!TextUtils.isEmpty(mItem) && !TextUtils.isEmpty(mState)) {
+                        longUri = "openhabitem://?" + QUERY_PARAMETER_ITEM_NAME + "=" + mItem
+                                + "&" + QUERY_PARAMETER_STATE + "=" + mState
+                                + "&" + QUERY_PARAMETER_MAPPED_STATE + "=" + mMappedState
+                                + "&" + QUERY_PARAMETER_ITEM_LABEL + "=" + mLabel;
+                        shortUri = "openhabitem://?" + QUERY_PARAMETER_ITEM_NAME + "=" + mItem
+                                + "&" + QUERY_PARAMETER_STATE + "=" + mState;
```

Comment:
Wouldn't it be more elegant to use the `android.net.Uri` class methods for constructing those? Could construct `shortUri` first and then `longUri` from it using `Uri.buildUpon()`.
That API should also work for `sitemapUri`.